fastpair/rust/demo/rust/src/advertisement.rs

```diff
@@ -0,0 +1,168 @@
+            "Fast Pair advertisements should advertise their transmit power."
+        ))?;
+
+        let distance = distance_from_rssi_and_tx_power(rssi, tx_power);
+
+        // Extract model ID from service data. We don't need to store service
+        // data in the `FpPairingAdvertisement` since it's easily accessible from
+        // `FpPairingAdvertisement.inner`, but it's convenient to save the parsed
+        // model ID.
+        let mut model_id =
+            FpDecoder::get_model_id_from_service_data(service_data).or_else(|err| {
+                // Some FP advertisements can be GATT non-discoverable
+                // advertisements containing service data that isn't
+                // the device model ID. In this case, simply ignore
+                // advertisements with errors extracting the model ID.
+                // See: developers.google.com/nearby/fast-pair/specifications/service/provider
+                Err(anyhow::anyhow!("error extracting model ID: {}", err))
+            })?;
+
+        if model_id.len() != 3 {
+            // In this demo of Fast Pair Rust, only model ID's
+            // of length 3 bytes are supported. Therefore, if a
+            // larger model ID makes it this far, log an error.
+            // TODO b/294453912
+            return Err(anyhow::anyhow!("Error: model ID of unsupported length"));
+        }
+
+        // Pad with 0 at the beginning to successfully call `from_be_bytes`.
+        // Assumes `model_id.len() == 3` before the call to `insert`, otherwise
+        // this will panic.
+        model_id.insert(0, 0);
+        let model_id = format!("{}", u32::from_be_bytes(model_id.try_into().unwrap()));
+
+        // Retrieve device info of the device corresponding to this model ID.
+        let device_info = FpDecoder::get_device_info_from_model_id(&model_id)
+            .expect("Failed to create device info from model ID.");
+
+        Ok(FpPairingAdvertisement {
+            inner: adv,
+            distance,
+            model_id,
+            name: device_info.name().to_string(),
+            image_url: device_info.image_url().to_string(),
+        })
+    }
+
+    /// Retrieve estimated distance the BLE advertisement travelled between
+    /// the sending device and this receiver.
+    pub(crate) fn distance(&self) -> f64 {
+        self.distance
+    }
+
+    /// Retrieve the BLE Address of the advertising device.
+    pub(crate) fn address(&self) -> BleAddress {
+        self.inner.address()
+    }
+
+    /// Retrieve the Model ID advertised by this device, parsed from the
+    /// 16-bit UUID service data.
+    pub(crate) fn model_id(&self) -> &ModelId {
+        &self.model_id
+    }
+
+    pub(crate) fn name(&self) -> &String {
+        &self.name
+    }
+
+    pub(crate) fn image_url(&self) -> &String {
+        &self.image_url
+    }
+}
+
+// Implement necessary Rust constructs to easily compute the closest device.
+impl PartialEq for FpPairingAdvertisement {
```

Comment:
This is a rather surprising definition of PartialEq. Same for PartialOrd. If you need to sort, use `sort_by_key` etc to be able to sort by distance if that's what you need in a particular circumstance.